File: services/media-service/internal/service/service.go
package service

import (
	"context"
	"errors"
	"fmt"
	"math/rand"
	"time"

	"github.com/google/uuid"
	"github.com/rs/zerolog/log"

	"github.com/elevatecompact/spark/services/media-service/internal/domain"
	"github.com/elevatecompact/spark/services/media-service/internal/repository"
)

type MediaService struct {
	repo *repository.MediaRepository
	evt  domain.EventProducer
}

func NewMediaService(repo *repository.MediaRepository, evt domain.EventProducer) *MediaService {
	return &MediaService{repo: repo, evt: evt}
}

func (s *MediaService) InitUpload(ctx context.Context, uploaderID uuid.UUID, filename, contentType string, fileSize int64) (*domain.UploadSession, error) {
	chunkSize := int64(5 * 1024 * 1024)
	chunksTotal := int(fileSize / chunkSize)
	if fileSize%chunkSize != 0 {
		chunksTotal++
	}
	session := &domain.UploadSession{
		ID:            uuid.New(),
		UploaderID:    uploaderID,
		Filename:      filename,
		FileSizeBytes: fileSize,
		ContentType:   contentType,
		ChunksTotal:   chunksTotal,
		ChunksDone:    0,
		Status:        "initiated",
		StoragePath:   fmt.Sprintf("uploads/%s/%s", uploaderID.String(), uuid.New().String()),
		CreatedAt:     time.Now(),
		ExpiresAt:     time.Now().Add(24 * time.Hour),
	}
	if err := s.repo.SaveUploadSession(ctx, session); err != nil {
		return nil, err
	}
	return session, nil
}

func (s *MediaService) UploadChunk(ctx context.Context, sessionID uuid.UUID, chunkIndex int) error {
	session, err := s.repo.GetUploadSession(ctx, sessionID)
	if err != nil {
		return err
	}
	if session.Status == "completed" {
		return errors.New("upload already completed")
	}
	done := session.ChunksDone + 1
	status := "uploading"
	if done >= session.ChunksTotal {
		status = "completed"
	}
	return s.repo.UpdateUploadSession(ctx, sessionID, done, "", status)
}

func (s *MediaService) CompleteUpload(ctx context.Context, sessionID uuid.UUID) (*domain.MediaAsset, error) {
	session, err := s.repo.GetUploadSession(ctx, sessionID)
	if err != nil {
		return nil, err
	}
	if session.ChunksDone < session.ChunksTotal {
		return nil, errors.New("upload not complete")
	}
	asset := &domain.MediaAsset{
		ID:             uuid.New(),
		UploaderID:     session.UploaderID,
		ContentType:    classifyContentType(session.ContentType),
		SourceFilename: session.Filename,
		FileSizeBytes:  session.FileSizeBytes,
		MimeType:       session.ContentType,
		Status:         domain.MediaStatusProcessing,
		StoragePath:    session.StoragePath,
		CDNURL:         fmt.Sprintf("https://cdn.spark.dev/media/%s", uuid.New().String()),
		DurationSecs:   0,
		Width:          1920,
		Height:         1080,
		Checksum:       session.Checksum,
		CreatedAt:      time.Now(),
	}
	if err := s.repo.CreateMediaAsset(ctx, asset); err != nil {
		return nil, err
	}
	s.repo.DeleteUploadSession(ctx, sessionID)
	if err := s.StartTranscoding(ctx, asset.ID); err != nil {
		log.Error().Err(err).Str("mediaId", asset.ID.String()).Msg("failed to start transcoding")
	}
	s.evt.Publish(ctx, "media.upload.completed", map[string]interface{}{
		"mediaId": asset.ID, "uploaderId": asset.UploaderID, "fileSize": asset.FileSizeBytes,
	})
	return asset, nil
}

func (s *MediaService) GetUploadStatus(ctx context.Context, sessionID uuid.UUID) (*domain.UploadSession, error) {
	return s.repo.GetUploadSession(ctx, sessionID)
}

func (s *MediaService) CancelUpload(ctx context.Context, sessionID uuid.UUID) error {
	return s.repo.DeleteUploadSession(ctx, sessionID)
}

func (s *MediaService) StartTranscoding(ctx context.Context, mediaID uuid.UUID) error {
	media, err := s.repo.GetMediaAsset(ctx, mediaID)
	if err != nil {
		return err
	}
	if media.ContentType != domain.ContentTypeVideo {
		return nil
	}
	profiles := []domain.RenditionProfile{domain.Rendition720p, domain.Rendition1080p, domain.RenditionSource}
	job := &domain.TranscodingJob{
		ID:        uuid.New(),
		MediaID:   mediaID,
		Profiles:  profiles,
		Status:    domain.TranscodingPending,
		CreatedAt: time.Now(),
	}
	if err := s.repo.CreateTranscodingJob(ctx, job); err != nil {
		return err
	}
	s.evt.Publish(ctx, "media.transcoding.started", map[string]interface{}{
		"mediaId": mediaID, "jobId": job.ID, "profiles": profiles,
	})
	go s.processTranscodingJob(context.Background(), job.ID)
	return nil
}

func (s *MediaService) processTranscodingJob(ctx context.Context, jobID uuid.UUID) {
	s.repo.UpdateTranscodingJobStatus(ctx, jobID, domain.TranscodingProcessing, "")
	time.Sleep(500 * time.Millisecond)
	job, err := s.repo.GetTranscodingJob(ctx, jobID)
	if err != nil {
		log.Error().Err(err).Str("jobId", jobID.String()).Msg("failed to get transcoding job")
		return
	}
	for _, profile := range job.Profiles {
		format := s.selectFormat(profile)
		cdn := fmt.Sprintf("https://cdn.spark.dev/media/%s/%s", job.MediaID.String(), string(profile))
		rd := &domain.MediaRendition{
			ID:      uuid.New(),
			MediaID: job.MediaID,
			Profile: profile,
			Format:  format,
			FileSizeBytes: int64(rand.Intn(50000000) + 1000000),
			StoragePath: fmt.Sprintf("renditions/%s/%s", job.MediaID.String(), string(profile)),
			CDNURL:  cdn,
			Status:  domain.MediaStatusReady,
			CreatedAt: time.Now(),
		}
		s.repo.CreateRendition(ctx, rd)
	}
	s.repo.UpdateTranscodingJobStatus(ctx, jobID, domain.TranscodingCompleted, "")
	s.repo.UpdateMediaCDN(ctx, job.MediaID, fmt.Sprintf("https://cdn.spark.dev/media/%s/playlist.m3u8", job.MediaID.String()))
	s.evt.Publish(ctx, "media.transcoding.completed", map[string]interface{}{
		"mediaId": job.MediaID, "jobId": job.ID,
		"profiles": job.Profiles,
	})
}

func (s *MediaService) selectFormat(profile domain.RenditionProfile) domain.RenditionFormat {
	switch profile {
	case domain.RenditionThumbnail:
		return domain.RenditionJPG
	case domain.Rendition720p, domain.Rendition1080p:
		return domain.RenditionHLS
	default:
		return domain.RenditionMP4
	}
}

func (s *MediaService) GenerateThumbnail(ctx context.Context, mediaID uuid.UUID, timeSecs float64) (*domain.MediaRendition, error) {
	media, err := s.repo.GetMediaAsset(ctx, mediaID)
	if err != nil {
		return nil, err
	}
	if media.ContentType != domain.ContentTypeVideo {
		return nil, errors.New("thumbnails only supported for video")
	}
	rd := &domain.MediaRendition{
		ID:      uuid.New(),
		MediaID: mediaID,
		Profile: domain.RenditionThumbnail,
		Format:  domain.RenditionJPG,
		FileSizeBytes: int64(rand.Intn(50000) + 10000),
		StoragePath: fmt.Sprintf("thumbnails/%s/%f.jpg", mediaID.String(), timeSecs),
		CDNURL:  fmt.Sprintf("https://cdn.spark.dev/media/%s/thumb/%f.jpg", mediaID.String(), timeSecs),
		Status:  domain.MediaStatusReady,
		CreatedAt: time.Now(),
	}
	if err := s.repo.CreateRendition(ctx, rd); err != nil {
		return nil, err
	}
	s.evt.Publish(ctx, "media.thumbnail.generated", map[string]interface{}{
		"mediaId": mediaID, "time": timeSecs, "url": rd.CDNURL,
	})
	return rd, nil
}

func (s *MediaService) OptimizeImage(ctx context.Context, mediaID uuid.UUID, width, height int) (*domain.MediaRendition, error) {
	media, err := s.repo.GetMediaAsset(ctx, mediaID)
	if err != nil {
		return nil, err
	}
	if media.ContentType != domain.ContentTypeImage {
		return nil, errors.New("optimization only supported for images")
	}
	format := domain.RenditionWebP
	if width == 0 {
		width = media.Width
	}
	if height == 0 {
		height = media.Height
	}
	rd := &domain.MediaRendition{
		ID:      uuid.New(),
		MediaID: mediaID,
		Profile: domain.RenditionProfile(fmt.Sprintf("%dx%d", width, height)),
		Format:  format,
		FileSizeBytes: int64(rand.Intn(100000) + 10000),
		StoragePath: fmt.Sprintf("optimized/%s/%dx%d.webp", mediaID.String(), width, height),
		CDNURL:  fmt.Sprintf("https://cdn.spark.dev/media/%s/opt/%dx%d.webp", mediaID.String(), width, height),
		Status:  domain.MediaStatusReady,
		CreatedAt: time.Now(),
	}
	if err := s.repo.CreateRendition(ctx, rd); err != nil {
		return nil, err
	}
	return rd, nil
}

func (s *MediaService) GetMediaInfo(ctx context.Context, mediaID uuid.UUID) (*domain.MediaAsset, error) {
	return s.repo.GetMediaAsset(ctx, mediaID)
}

func (s *MediaService) GetMediaStatus(ctx context.Context, mediaID uuid.UUID) (*domain.MediaAsset, error) {
	return s.repo.GetMediaAsset(ctx, mediaID)
}

func (s *MediaService) GetPlaybackURL(ctx context.Context, mediaID uuid.UUID) (string, error) {
	media, err := s.repo.GetMediaAsset(ctx, mediaID)
	if err != nil {
		return "", err
	}
	if media.Status != domain.MediaStatusReady {
		return "", errors.New("media not ready")
	}
	return fmt.Sprintf("https://cdn.spark.dev/media/%s/playlist.m3u8", mediaID.String()), nil
}

func (s *MediaService) GetDownloadURL(ctx context.Context, mediaID uuid.UUID) (string, error) {
	_, err := s.repo.GetMediaAsset(ctx, mediaID)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("https://cdn.spark.dev/media/%s/download", mediaID.String()), nil
}

func (s *MediaService) GetThumbnailURL(ctx context.Context, mediaID uuid.UUID, timeSecs float64) (string, error) {
	return fmt.Sprintf("https://cdn.spark.dev/media/%s/thumb/%f.jpg", mediaID.String(), timeSecs), nil
}

func (s *MediaService) IssueDRMLicense(ctx context.Context, mediaID uuid.UUID, keySystem string) (string, error) {
	s.evt.Publish(ctx, "media.drm.license.issued", map[string]interface{}{
		"mediaId": mediaID, "keySystem": keySystem,
	})
	return fmt.Sprintf("drm-license-%s-%s", mediaID.String(), uuid.New().String()), nil
}

func (s *MediaService) CreateDRMPolicy(ctx context.Context, p *domain.DRMPolicy) (*domain.DRMPolicy, error) {
	p.ID = uuid.New()
	p.CreatedAt = time.Now()
	if err := s.repo.CreateDRMPolicy(ctx, p); err != nil {
		return nil, err
	}
	return p, nil
}

func (s *MediaService) GetDRMPolicies(ctx context.Context) ([]domain.DRMPolicy, error) {
	return s.repo.GetDRMPolicies(ctx)
}

func (s *MediaService) DeleteMedia(ctx context.Context, mediaID uuid.UUID) error {
	if err := s.repo.DeleteMediaAsset(ctx, mediaID); err != nil {
		return err
	}
	s.evt.Publish(ctx, "media.content.deleted", map[string]interface{}{
		"mediaId": mediaID,
	})
	return nil
}

func (s *MediaService) GetStorageUsage(ctx context.Context) (*domain.StorageUsage, error) {
	return s.repo.GetStorageUsage(ctx)
}

func (s *MediaService) PurgeCDN(ctx context.Context, path string) error {
	log.Info().Str("path", path).Msg("CDN cache purged (noop)")
	return nil
}

func (s *MediaService) GetProcessingQueue(ctx context.Context, limit, offset int) ([]domain.TranscodingJob, error) {
	status := domain.TranscodingPending
	return s.repo.ListTranscodingJobs(ctx, &status, limit, offset)
}

func (s *MediaService) RetryTranscoding(ctx context.Context, mediaID uuid.UUID) error {
	return s.StartTranscoding(ctx, mediaID)
}

func (s *MediaService) GetRenditions(ctx context.Context, mediaID uuid.UUID) ([]domain.MediaRendition, error) {
	return s.repo.GetRenditions(ctx, mediaID)
}

func classifyContentType(mime string) domain.ContentType {
	switch {
	case mime == "video/mp4" || mime == "video/webm" || mime == "video/quicktime":
		return domain.ContentTypeVideo
	case mime == "image/jpeg" || mime == "image/png" || mime == "image/webp" || mime == "image/gif":
		return domain.ContentTypeImage
	case mime == "audio/mpeg" || mime == "audio/ogg" || mime == "audio/wav":
		return domain.ContentTypeAudio
	}
	return domain.ContentTypeVideo
}

func (s *MediaService) GetMediaByUploader(ctx context.Context, uploaderID uuid.UUID, limit, offset int) ([]domain.MediaAsset, error) {
	return s.repo.GetMediaByUploader(ctx, uploaderID, limit, offset)
}

func (s *MediaService) GetTranscodingJob(ctx context.Context, jobID uuid.UUID) (*domain.TranscodingJob, error) {
	return s.repo.GetTranscodingJob(ctx, jobID)
}

func (s *MediaService) IsMediaReady(ctx context.Context, mediaID uuid.UUID) (bool, error) {
	media, err := s.repo.GetMediaAsset(ctx, mediaID)
	if err != nil {
		return false, err
	}
	return media.Status == domain.MediaStatusReady, nil
}

File: services/media-service/src/service.rs
use std::sync::Arc;
use chrono::{Duration, Utc};
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use tokio::time::sleep;
use uuid::Uuid;

use crate::domain::*;
use crate::events::EventProducer;
use crate::repository::MediaRepository;

pub struct MediaService {
    repo: Arc<MediaRepository>,
    pub evt: Arc<dyn EventProducer + Send + Sync>,
}

impl MediaService {
    pub fn new(repo: Arc<MediaRepository>, evt: Arc<dyn EventProducer + Send + Sync>) -> Self {
        Self { repo, evt }
    }

    pub async fn init_upload(&self, req: &InitUploadRequest) -> Result<UploadSession, String> {
        let chunk_size: i64 = 5 * 1024 * 1024;
        let chunks_total = ((req.file_size_bytes + chunk_size - 1) / chunk_size) as i32;
        let session = UploadSession {
            id: Uuid::new_v4(),
            uploader_id: req.uploader_id,
            filename: req.filename.clone(),
            file_size_bytes: req.file_size_bytes,
            content_type: req.content_type.clone(),
            chunks_total,
            chunks_done: 0,
            checksum: String::new(),
            status: "initiated".into(),
            storage_path: format!("uploads/{}/{}", req.uploader_id, Uuid::new_v4()),
            created_at: Utc::now(),
            expires_at: Utc::now() + Duration::hours(24),
        };
        self.repo.save_upload_session(&session).await.map_err(|e| e.to_string())?;
        Ok(session)
    }

    pub async fn upload_chunk(&self, session_id: Uuid) -> Result<(), String> {
        let session = self.repo.get_upload_session(session_id).await.map_err(|e| e.to_string())?;
        if session.status == "completed" {
            return Err("upload already completed".into());
        }
        let done = session.chunks_done + 1;
        let status = if done >= session.chunks_total { "completed" } else { "uploading" };
        self.repo.update_upload_session(session_id, done, status).await.map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn complete_upload(&self, session_id: Uuid) -> Result<MediaAsset, String> {
        let session = self.repo.get_upload_session(session_id).await.map_err(|e| e.to_string())?;
        if session.chunks_done < session.chunks_total {
            return Err("upload not complete".into());
        }
        let asset = MediaAsset {
            id: Uuid::new_v4(),
            uploader_id: session.uploader_id,
            content_type: classify_content_type(&session.content_type),
            source_filename: session.filename.clone(),
            file_size_bytes: session.file_size_bytes,
            mime_type: session.content_type.clone(),
            status: MediaStatus::Processing,
            storage_path: session.storage_path.clone(),
            cdn_url: format!("https://cdn.spark.dev/media/{}", Uuid::new_v4()),
            duration_seconds: 0.0,
            width: 1920,
            height: 1080,
            checksum: session.checksum.clone(),
            created_at: Utc::now(),
        };
        self.repo.create_media_asset(&asset).await.map_err(|e| e.to_string())?;
        self.repo.delete_upload_session(session_id).await.ok();
        let _ = self.start_transcoding(asset.id).await;
        self.evt.publish("media.upload.completed", serde_json::json!({
            "mediaId": asset.id, "uploaderId": asset.uploader_id, "fileSize": asset.file_size_bytes
        })).await;
        Ok(asset)
    }

    pub async fn get_upload_status(&self, session_id: Uuid) -> Result<UploadSession, String> {
        self.repo.get_upload_session(session_id).await.map_err(|e| e.to_string())
    }

    pub async fn cancel_upload(&self, session_id: Uuid) -> Result<(), String> {
        self.repo.delete_upload_session(session_id).await.map_err(|e| e.to_string())
    }

    pub async fn start_transcoding(&self, media_id: Uuid) -> Result<(), String> {
        let media = self.repo.get_media_asset(media_id).await.map_err(|e| e.to_string())?;
        if media.content_type != ContentType::Video {
            return Ok(());
        }
        let profiles = vec!["720p".into(), "1080p".into(), "source".into()];
        let job = TranscodingJob {
            id: Uuid::new_v4(),
            media_id,
            profiles: profiles.clone(),
            status: TranscodingStatus::Pending,
            worker_id: String::new(),
            started_at: None,
            completed_at: None,
            error_message: String::new(),
            created_at: Utc::now(),
        };
        self.repo.create_transcoding_job(&job).await.map_err(|e| e.to_string())?;
        self.evt.publish("media.transcoding.started", serde_json::json!({
            "mediaId": media_id, "jobId": job.id, "profiles": profiles
        })).await;
        let repo = self.repo.clone();
        let evt = self.evt.clone();
        tokio::spawn(async move {
            process_transcoding(repo, evt, job.id).await;
        });
        Ok(())
    }

    pub async fn get_media_status(&self, media_id: Uuid) -> Result<MediaAsset, String> {
        self.repo.get_media_asset(media_id).await.map_err(|e| e.to_string())
    }

    pub async fn get_media_info(&self, media_id: Uuid) -> Result<MediaAsset, String> {
        self.repo.get_media_asset(media_id).await.map_err(|e| e.to_string())
    }

    pub async fn get_playback_url(&self, media_id: Uuid) -> Result<String, String> {
        let media = self.repo.get_media_asset(media_id).await.map_err(|e| e.to_string())?;
        if media.status != MediaStatus::Ready {
            return Err("media not ready".into());
        }
        Ok(format!("https://cdn.spark.dev/media/{}/playlist.m3u8", media_id))
    }

    pub async fn get_download_url(&self, media_id: Uuid) -> Result<String, String> {
        Ok(format!("https://cdn.spark.dev/media/{}/download", media_id))
    }

    pub async fn get_thumbnail_url(&self, media_id: Uuid, time_secs: f64) -> Result<String, String> {
        Ok(format!("https://cdn.spark.dev/media/{}/thumb/{}.jpg", media_id, time_secs))
    }

    pub async fn generate_thumbnail(&self, req: &ThumbnailRequest) -> Result<MediaRendition, String> {
        let mut rng = StdRng::from_entropy();
        let rendition = MediaRendition {
            id: Uuid::new_v4(),
            media_id: req.media_id,
            profile: "thumbnail".into(),
            format: "jpg".into(),
            file_size_bytes: rng.gen_range(10000..50000),
            storage_path: format!("thumbnails/{}/{}.jpg", req.media_id, req.time_seconds),
            cdn_url: format!("https://cdn.spark.dev/media/{}/thumb/{}.jpg", req.media_id, req.time_seconds),
            status: MediaStatus::Ready,
            created_at: Utc::now(),
        };
        self.repo.create_rendition(&rendition).await.map_err(|e| e.to_string())?;
        self.evt.publish("media.thumbnail.generated", serde_json::json!({
            "mediaId": req.media_id, "time": req.time_seconds, "url": rendition.cdn_url
        })).await;
        Ok(rendition)
    }

    pub async fn optimize_image(&self, req: &OptimizeRequest) -> Result<MediaRendition, String> {
        let media = self.repo.get_media_asset(req.media_id).await.map_err(|e| e.to_string())?;
        if media.content_type != ContentType::Image {
            return Err("optimization only supported for images".into());
        }
        let mut rng = StdRng::from_entropy();
        let rendition = MediaRendition {
            id: Uuid::new_v4(),
            media_id: req.media_id,
            profile: format!("{}x{}", req.width, req.height),
            format: "webp".into(),
            file_size_bytes: rng.gen_range(10000..100000),
            storage_path: format!("optimized/{}/{}x{}.webp", req.media_id, req.width, req.height),
            cdn_url: format!("https://cdn.spark.dev/media/{}/opt/{}x{}.webp", req.media_id, req.width, req.height),
            status: MediaStatus::Ready,
            created_at: Utc::now(),
        };
        self.repo.create_rendition(&rendition).await.map_err(|e| e.to_string())?;
        Ok(rendition)
    }

    pub async fn get_renditions(&self, media_id: Uuid) -> Result<Vec<MediaRendition>, String> {
        self.repo.get_renditions(media_id).await.map_err(|e| e.to_string())
    }

    pub async fn issue_drm_license(&self, req: &DrmLicenseRequest) -> Result<String, String> {
        self.evt.publish("media.drm.license.issued", serde_json::json!({
            "mediaId": req.media_id, "keySystem": req.key_system
        })).await;
        Ok(format!("drm-license-{}-{}", req.media_id, Uuid::new_v4()))
    }

    pub async fn create_drm_policy(&self, p: &DrmPolicy) -> Result<DrmPolicy, String> {
        let mut policy = p.clone();
        policy.id = Uuid::new_v4();
        policy.created_at = Utc::now();
        self.repo.create_drm_policy(&policy).await.map_err(|e| e.to_string())?;
        Ok(policy)
    }

    pub async fn get_drm_policies(&self) -> Result<Vec<DrmPolicy>, String> {
        self.repo.get_drm_policies().await.map_err(|e| e.to_string())
    }

    pub async fn get_storage_usage(&self) -> Result<StorageUsage, String> {
        self.repo.get_storage_usage().await.map_err(|e| e.to_string())
    }

    pub async fn purge_cdn(&self, path: &str) -> Result<(), String> {
        tracing::info!(path, "CDN cache purged (noop)");
        Ok(())
    }

    pub async fn get_processing_queue(&self, limit: i64, offset: i64) -> Result<Vec<TranscodingJob>, String> {
        self.repo.list_transcoding_jobs(Some("pending"), limit, offset).await.map_err(|e| e.to_string())
    }

    pub async fn retry_transcoding(&self, media_id: Uuid) -> Result<(), String> {
        self.start_transcoding(media_id).await
    }
}

async fn process_transcoding(repo: Arc<MediaRepository>, evt: Arc<dyn EventProducer + Send + Sync>, job_id: Uuid) {
    let _ = repo.update_transcoding_job_status(job_id, &TranscodingStatus::Processing, "").await;
    sleep(std::time::Duration::from_millis(500)).await;
    let job = repo.get_transcoding_job(job_id).await;
    if let Ok(job) = job {
    let mut rng = StdRng::from_entropy();
    for profile in &job.profiles {
            let format = if profile == "thumbnail" { "jpg" } else { "hls" };
            let rendition = MediaRendition {
                id: Uuid::new_v4(),
                media_id: job.media_id,
                profile: profile.clone(),
                format: format.into(),
                file_size_bytes: rng.gen_range(1000000..50000000),
                storage_path: format!("renditions/{}/{}", job.media_id, profile),
                cdn_url: format!("https://cdn.spark.dev/media/{}/{}", job.media_id, profile),
                status: MediaStatus::Ready,
                created_at: Utc::now(),
            };
            let _ = repo.create_rendition(&rendition).await;
        }
        let _ = repo.update_transcoding_job_status(job_id, &TranscodingStatus::Completed, "").await;
        let _ = repo.update_media_cdn(job.media_id, &format!("https://cdn.spark.dev/media/{}/playlist.m3u8", job.media_id)).await;
        evt.publish("media.transcoding.completed", serde_json::json!({
            "mediaId": job.media_id, "jobId": job.id, "profiles": job.profiles
        })).await;
    }
}

fn classify_content_type(mime: &str) -> ContentType {
    match mime {
        "video/mp4" | "video/webm" | "video/quicktime" => ContentType::Video,
        "image/jpeg" | "image/png" | "image/webp" | "image/gif" => ContentType::Image,
        "audio/mpeg" | "audio/ogg" | "audio/wav" => ContentType::Audio,
        _ => ContentType::Video,
    }
}
